### ISSUE TITLE
Refactor: BlockTemplatesController.php file

### ISSUE THREAD

**Issue by tjcafferkey** [woothemes/woocommerce#42251]
## Is your feature request related to a problem? Please describe.
Currently the [`BlockTemplatesController.php`](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/BlockTemplatesController.php) is responsible for critical features and houses some very important logic.

In its current state the controller is extremely complex, hard to reason with and we are continuing to add logic to this class and in doing some we are just kicking the can down the road as we incur more and more technical debt in here.

## Describe the solution you'd like
I would like to see this controller refactored into a more readable and robust solution, however that may look. We likely need to do an audit of its current state, its responsibilities and edge cases it is caters for before determining the best path forward.

## Describe alternatives you've considered
N/A

## Additional context
We do have some E2E tests around functionality implemented in this class, hopefully this will serve us well during a refactor.

```[tasklist]
### Task
- [ ] https://github.com/woocommerce/woocommerce/issues/42191
- [ ] https://github.com/woocommerce/woocommerce/issues/42181
```

**Comment by gigitux** [comment]
Some tasks that we have to do for consider the refactor successful:

- [ ] Add some coverage with unit tests/E2E tests.
- [ ] Make the class smaller and more readable. 
- [ ] Refactor the [`render_block_template`](https://github.com/woocommerce/woocommerce-blocks/blob/5b83f99a2bb01608e3900d6c59b904f1d672c810/src/BlockTemplatesController.php/#L511-L594) function is complex to debug/understand.
- [ ] Not use `_build_block_template_result_from_post` private function. ([source code](https://github.com/woocommerce/woocommerce-blocks/blob/2b79d3e87dffd5b38d5910ebcf39057e267d190d/src/BlockTemplatesController.php/#L118-L119)).
- [ ] For the Single Product template, some classes are added via `get_block_templates`. This is not the appropriate place and causes some odd results. Check woocommerce/woocommerce-blocks#9697 for more details.

(copy-pasted from woocommerce/woocommerce-blocks#9491)

**Comment by gigitux** [comment]
I converted this issue into an epic 👍

**Comment by Aljullu** [comment]
Part of the refactor work is done in https://github.com/woocommerce/woocommerce/pull/44537.

As mentioned in the comments of that PR (see https://github.com/woocommerce/woocommerce/pull/44537#pullrequestreview-1882228400 for example), we would still like to work on refactoring all the logic related to the "fallback system" between templates in a follow-up PR.

Note: it might make sense to work on this in parallel or after https://github.com/woocommerce/woocommerce/issues/42531 and https://github.com/woocommerce/woocommerce/issues/42532. As those PRs will change significantly how the fallback system is implemented.

**Comment by Aljullu** [comment]
I'm going to close this issue in favor of https://github.com/woocommerce/woocommerce/issues/47885. While we still have several open issues & PRs about improving `BlockTemplatesController`, I think the next big step should be using the upstream API.
